### PR TITLE
Raspberry Pi OS Bookworm + Pep 668

### PR DESCRIPTION
**Pull request by Gadgetoid**
⚠️  Progress / TODO List moved to https://github.com/pimoroni/boilerplate-python/issues/16

# Rationale

Update:

* I have switched to `~/.virtualenvs/pimoroni' as the default virtualenv dir, since this is what `virtualenvwrapper` uses, and thus is one of the search paths used by Visual Studio Code.
* I have raised a PR with Thonny to get this same directory supported there, so we can have some cohesion between CLI, VSCode, Thonny - https://github.com/thonny/thonny/pull/2978

As of RPi OS Bookworm (required for the Raspberry Pi 5) incorporating [Pep 668](https://peps.python.org/pep-0668/), we are required to install Python packages into a virtual environment or "venv".

The TLDR of 668 is twofold -

* Allowing users to affect the system-wide Python installation could lead to broken system packages or a broken system Python environment. Therefore we should not.
* Using a virtual environment allows dependencies to be managed per project, so the chance of conflicting dependencies (one library needs v1.0.0 another needs v2.0.0 for example) is reduced.

Attempting to modify the system Python environment will kick out an error along the lines of:

```
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
  python3-xyz, where xyz is the package you are trying to
  install.
  ```

[Raspberry Pi's own documentation](https://www.raspberrypi.com/documentation/computers/os.html#python-on-raspberry-pi) (linked in this error message on RPi OS) attempts to cover some basic use cases of virtual environments ans shows some per-project and per user patterns you might consider if you're keen to learn.

For most beginners, however, the specifics of setting up a virtual environment are largely meaningless and due to the non-destructive - you might say "virtual" - nature of these environments there is no harm in us stepping in as the arbiter of truth and setting up one by default.

We can do this in a couple of ways:

# Supporting venvs in our Python boilerplate

The changes in *this* PR attempt to detect and (optionally) manage a virtual environment for the user.

The virtual environment we use by default will live in `$HOME/.virtualenvs/pimoroni` and this will be shared across all packages, giving a single common mixed use venv for beginners.

# Providing a simple "manage me a venv" script

It's possible for a user to opt to gloss over the existence of a virtual environment by adding the following lines to `~/.bashrc` - 

```bash
PY_ENV_DIR=$HOME/.virtualenvs/pimoroni
if [ ! -f $PY_ENV_DIR/bin/activate ]; then
  printf "Creating user Python environment in $PY_ENV_DIR, please wait...\n"
  mkdir -p $PY_ENV_DIR
  python3 -m venv --system-site-packages $PY_ENV_DIR
fi
printf " ↓ ↓ ↓ ↓   Hello, we've activated a Python venv for you. To exit, type \"deactivate\".\n"
source $PY_ENV_DIR/bin/activate
```

This will create (if it does not exist) and activate a Python virtual environment automatically any time a bash shell is opened. If the user doesn't want to use it, they can type "deactivate" or switch into another virtual environment just as easily.

The idea behind this is that users who have installed our software, and are set up to use our `$HOME/.virtualenvs/pimoroni` virtual environment don't need to explicitly activate it every time they want to work with our products.

Putting this into `~/.bashrc` automatically would be unwise since:

* It's a little rude, since it's difficult to explain in crystal clear detail just how this will modify the users shell
* It's quite difficult to check if it's already there
* It's difficult to automatically update, should it ever need changing
* We don't know what else is in `~/.bashrc` and how it might (negatively) interact with this snippet
* There are many instances where it doesn't help or take effect; ie: VSCode or Thonny.

But it's probably *useful* for users who really don't want to have to think about virtual environments (much.)

# System Packages

In both cases, since we still rely on apt packages to satisfy some dependencies, we're using `--system-site-packages` to pass through the system Python packages into the virtualenv. This is generally unwise since it introduces packages we can't properly manage into the virtualenv, but for now it permits the use of `gpiozero` + `lgpio` (the latter missing a proper pypi package) and 'libgpiod' (also currently missing a working pypi package),

# What about my precious rpi_ws281x?

In order to run Python code as root within the virtualenv context you need to:

```
sudo --preserve-env=PATH python my_code.py
```

# In conclusion

Many users don't really want to know or care about a per-project virtual Python environment, and just want a close approximation of the old behaviour. This gets us part of the way with the notable exceptions that:

* Running Python code as root is more complicated (see above)
* System-owned Python packages are made available to the virtual environment, but this may cause conflicts between older (RPi OS ships with a 3 year old libgpiod 1.6.3) system packages a newer Pypi packages.
* It's not always obvious how to use a virtual env with a particular IDE, Thonny's UX around venvs- for example - is, confusing.

### CODE REVIEW

**Comment by aallan** [comment]
> [Raspberry Pi's own documentation](https://www.raspberrypi.com/documentation/computers/os.html#python-on-raspberry-pi) (linked in this error message on RPi OS) attempts to cover some basic use cases of virtual environments ans shows some per-project and per user patterns you might consider if you're keen to learn.

If you have anything that should go into that documentation, which as you mention is linked right in the error message, please either drop me email or raise a PR against the `develop` branch on the [docs repo](https://github.com/raspberrypi/documentation).

**Comment by Gadgetoid** [comment]
> linked right in the error message

The two people who actually read error messages will appreciate that :sob:

Right now I don't have any grand designs on the documentation, my preference is mostly to smooth over the transition as much as possible via our installers (due to aforementioned and unfortunately not sarcastic problem with people simply not reading errors or documentation).

I will add that it did confuse me a little when I clicked through and went straight to "Python on Raspberry Pi" rather than "About Python virtual environments."

(Thanks for finding your way here, by the way, I assume you followed my whimpering cries of anguish)

File: install.sh [hunk 1]
```diff
@@ -62,6 +55,28 @@ warning() {
...
 
+venv_check() {
+	PYTHON_BIN=`which $PYTHON`
+	if [[ $VIRTUAL_ENV == "" ]] || [[ $PYTHON_BIN != $VIRTUAL_ENV* ]]; then
+		printf "This script should be run in a virtual Python environment.\n"
+		if confirm "Would you like us to create one for you?"; then
+			if [ ! -f $PY_VENV_DIR/bin/activate ]; then
+				inform "Creating virtual Python environment in $PY_VENV_DIR, please wait...\n"
+				mkdir -p $PY_VENV_DIR
+				/usr/bin/python3 -m venv $PY_VENV_DIR --system-site-packages
```
Comment (rglidden) [inline install.sh:66]: It might be nice to add `--prompt Pimoroni` or something similar to the creation of the venv so that when the environment is activated, the user's prompt changes to `(Pimoroni) pi@raspberrypi:~ $` which at least indicates that they're in a Pimoroni environment.  The default prompt of `(venv) pi@raspberrypi:~ $` is a little confusing as it's not clear _which_ "venv" you have activated.

Comment (Gadgetoid) [reply install.sh:66]: Thank you for taking the time to look at this. I'm keenly aware that foisting something like this upon users should be done at least *somewhat* carefully, and thus any and all feedback is really, really appreciated.

TLDR: Yes, I totally agree!

**Comment by Gadgetoid** [comment]
Considering recommending an extra line in `~/.bashrc` to make the terminal startup output something like:

```
 ↓ ↓ ↓ ↓   Hello, we've activated a Python venv for you. To exit, type "deactivate".
(Pimoroni) phil@ubuntu-pi5:~$
```

**Comment by aallan** [comment]
> > linked right in the error message
> 
> The two people who actually read error messages will appreciate that 😭

I mean, the folks who read it seem happy afterward, e.g. https://twitter.com/drfootleg/status/1714906054302060692.

> Right now I don't have any grand designs on the documentation, my preference is mostly to smooth over the transition as much as possible via our installers (due to aforementioned and unfortunately not sarcastic problem with people simply not reading errors or documentation).

I mean, I don't disagree.

> I will add that it did confuse me a little when I clicked through and went straight to "Python on Raspberry Pi" rather than "About Python virtual environments."

Future-proofing myself, I've got a sneaking suspicion that I'm going to need more verbiage eventually.
 
> (Thanks for finding your way here, by the way, I assume you followed my whimpering cries of anguish)

Pretty much! 😆

**Comment by aallan** [comment]
> Considering recommending an extra line in `~/.bashrc` to make the terminal startup output something like:
> 
> ```
>  ↓ ↓ ↓ ↓   Hello, we've activated a Python venv for you. To exit, type "deactivate".
> (Pimoroni) phil@ubuntu-pi5:~$
> ```

This sounds like a good plan!

**Comment by rbricheno** [comment]
This is a noble attempt at helping, which really shopuld be addressed at the OS level. I raised an issue here https://github.com/raspberrypi/bookworm-feedback/issues/100

**Comment by Footleg** [comment]
My attempts at following the suggestions here, on Bookworm 64bit OS on the Pi 5. First I tried to run the example without sudo (predicably it cannot access the LEDs). Then I try with sudo but still it fails.

```shell
(robots-venv) pi@raspberrypi5:~/inventorhatmini-python/examples $ python led_rainbow.py
Traceback (most recent call last):
  File "/home/pi/inventorhatmini-python/examples/led_rainbow.py", line 18, in <module>
    board = InventorHATMini()
            ^^^^^^^^^^^^^^^^^
  File "/home/pi/robots-venv/lib/python3.11/site-packages/inventorhatmini/__init__.py", line 99, in __init__
    GPIO.setup(self.PI_USER_SW_PIN, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
RuntimeError: No access to /dev/mem.  Try running as root!
Exception ignored in: <function InventorHATMini.__del__ at 0x7fff52ecd3a0>
Traceback (most recent call last):
  File "/home/pi/robots-venv/lib/python3.11/site-packages/inventorhatmini/__init__.py", line 142, in __del__
    self.ioe.reset()
    ^^^^^^^^
AttributeError: 'InventorHATMini' object has no attribute 'ioe'
(robots-venv) pi@raspberrypi5:~/inventorhatmini-python/examples $ sudo --preserve-env=PATH python led_rainbow.py
Traceback (most recent call last):
  File "/home/pi/inventorhatmini-python/examples/led_rainbow.py", line 18, in <module>
    board = InventorHATMini()
            ^^^^^^^^^^^^^^^^^
  File "/home/pi/robots-venv/lib/python3.11/site-packages/inventorhatmini/__init__.py", line 99, in __init__
    GPIO.setup(self.PI_USER_SW_PIN, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
RuntimeError: Mmap of GPIO registers failed
Exception ignored in: <function InventorHATMini.__del__ at 0x7fff9a65d3a0>
Traceback (most recent call last):
  File "/home/pi/robots-venv/lib/python3.11/site-packages/inventorhatmini/__init__.py", line 142, in __del__
    self.ioe.reset()
    ^^^^^^^^
AttributeError: 'InventorHATMini' object has no attribute 'ioe'
```

**Comment by Gadgetoid** [comment]
Yup RPi.GPIO - which any of our drivers using GPIO will use - is pretty much dead in the water at the moment. That's why I'm embroiled in getting the libgpiod Python bindings ship shape, so I can start migrating drivers to those.

There is an lgpio (which uses libgpiod) based compatibility shim for RPi.GPIO which might help: https://github.com/waveform80/rpi-lgpio

I've been too mired in getting the absolute basics working to see if there's any short term crutch we can rely on, however.

**Comment by ahnlak** [comment]
I raised an issue for the Grow HAT (https://github.com/pimoroni/grow-python/issues/38), but it's probably worth flagging here because I suspect it may not be the only case - part of the install there sets up a (python) service to run (as root, presumably) on startup, and I'm not sure what the 'right' way of making things like that understand the new `venv` world?

(I did try a couple of times but everything ended up horribly broken and I gave up, hoping someone else had thought of something...!)

**Comment by Gadgetoid** [comment]
>  I'm not sure what the 'right' way of making things like that understand the new `venv` world?

That makes two of us!

I suspect both systemd and the more humble crontab probably face similar issues, though in the latter case I think it runs in the user context so a simple launcher to invoke the right virtual environment before running the Python script might suffice?

One might argue that as soon as you're launching Grow, or a similar service, via systemd then it should be (along with its dependencies 💀) installed and managed by apt as a system dependency. I don't much like that argument because it means painful packaging practices potential loom in my future.

Any other solution would, I suppose, be *not* the right way... though not materially different from the concept of a Flatpak in theory so even that's debatable.

Do you remember what you tried?

**Comment by ahnlak** [comment]
I think I was just trying to launch it via `~/.virtualenvs/pimoroni/bin/python3`, but it was throwing up a lot of missing packages because I think it wasn't going a very good job of grabbing the `venv` version of packages, for some reason.

tbh after I worked out how to regress things to the point where a Bullseye install worked, my motivation tailed of somewhat...

**Comment by Gadgetoid** [comment]
From what I can dig up elsewhere, it looks like the Python interpreter in the venv should already handle PYTHONPATH (ie: where to find its installed libraries) so your approach *should* work. Albeit it seems there are some (probably not applicable to us) caveats- https://stackoverflow.com/questions/37211115/how-to-enable-a-virtualenv-in-a-systemd-service-unit

Strikes me that having a single Pimoroni virtual environment is probably a hit and miss idea when it comes to system services, ideally you'd want the service and its dependencies to be completely self-contained so you don't accidentally mess them up tinkering with something else. That's a small concession though, I'd guess, we can prompt the user "Do you want an isolated environment for this system service," drop a new venv, activate it and re-run the package installer.

File: install.sh [hunk 2]
```diff
@@ -56,6 +57,22 @@ warning() {
...
+find_config() {
+	if [ ! -f "$CONFIG_DIR/$CONFIG_FILE" ]; then
+		CONFIG_DIR="/boot"
+		if [ ! -f "$CONFIG_DIR/$CONFIG_FILE"]; then
+			warning "Could not find $CONFIG_FILE!"
+			exit 1
+		fi
+    else
+        if [ -f "/boot/$CONFIG_FILE" ] && [ ! -L "/boot/$CONFIG_FILE" ]; then
+            warning "Oops! It looks like /boot/$CONFIG_FILE is not a link to $CONFIG_DIR/$CONFIG_FILE"
```
Comment (Gadgetoid) [inline install.sh]: My latest Pi OS install has a non-link file for `/boot/config.txt` with the contents:

```
DO NOT EDIT THIS FILE

The file you are looking for has moved to /boot/firmware/config.txt
```

Need to check if this is now standard across all Pi OS installs and, in any case, check for this text and avoid emitting this warning.

A simpler path might be to avoid all this kerfuffle and always prefer `/boot/firmware/config.txt` if it exists, without any efforts to warn about other config files.

Or maybe even fail outright if `/boot/firmware/config.txt` does not exist...

Comment (Gadgetoid) [reply install.sh]: Okay I found the change: https://github.com/RPi-Distro/pi-gen/commit/a12534473182eb234a1751da7348f784b9271f42

It *will* be standard, but a quick test of 64bit Pi OS lite suggests it currently is not. Therefore the symlink warning is obsolete.

Comment (bsimmo) [reply install.sh]: Just tried growhat on bookworn 32bit, and it blew up at me, this was one of the problems.  It doesn't have anything in /boot/firmware/ at all.

Comment (Gadgetoid) [reply install.sh]: Grow should have a fallback to `/boot/config.txt`- do you have a full output of the install script?

Comment (bsimmo) [reply install.sh]: I'll try again, but nothing was added to the normal config.txt.

For now to get it working I've moved to legacy 'bullseye' and the install was ok.

I'll have another go with bookworm when I get the time.   (Then for fun I'll try the EnviroHat and the WeatherHat and the led matrix pi zero sized thing if forget the name of... obviously not as much fun as you're having rewriting all of this for all the devices.

But the venv never seemed to go right either, it didn't boot to one. and then when I manually enabled it none of the modules seemed to be there when running the examples. I'll check what it actually contains properly, it might be a few days though.
 
(Pi Zero, so has to be the 32bit lite version).

Comment (Gadgetoid) [reply install.sh]: Currently I'm not enabling the virtual environment by default, though it should install a script into `~/Pimoroni/auto_venv.sh` which can be optionally added to `~/.bashrc`.

`source ~/.virtualenvs/pimoroni/bin/activate` should activate the venv manually.
